### PR TITLE
New: add regexp-match rule (fixes #2829)

### LINKED ISSUE #2829: handle regexp for variable names
Right now we have a Require Camelcase rule  that makes sur variable/property names are camel case.
Would it be possible to provide a regex to match those names against ?

### PR DESCRIPTION
**Pull request by mlarcher**
This rule allows regular expression matching on Identifiers.

### CODE REVIEW

**Comment by mlarcher** [comment]
This PR fixes https://github.com/eslint/eslint/issues/2829

File: tests/lib/rules/regexp-match.js [hunk 1]
```diff
@@ -0,0 +1,262 @@
...
+//------------------------------------------------------------------------------
+// Tests
+//------------------------------------------------------------------------------
+
+var eslintTester = new ESLintTester(eslint);
+eslintTester.addRuleTest("lib/rules/regexp-match", {
+    valid: [
+        {
+            code: "firstname = \"Matthieu\"",
+            args: [2, "^[a-z]+$"]
```
Comment (nzakas) [inline tests/lib/rules/regexp-match.js]: Use `options` instead of `args`. When you do so, you can leave off the initial 2, as it is not used.

**Comment by nzakas** [comment]
This looks really good, though we need to work on the name. Right now it sounds like it operates on regular expressions. Maybe `id-match`?

**Comment by mlarcher** [comment]
Sounds good to me :)
I've modified the PR to use options instead of args in the tests and change the name.

File: docs/rules/id-match.md [hunk 2]
```diff
@@ -0,0 +1,72 @@
+# Require IdMatch (id-match)
+
+"There are only two hard things in Computer Science: cache invalidation and naming things." � Phil Karlton
```
Comment (btmills) [inline docs/rules/id-match.md]: Inserting a ">" at the beginning of this line and a blank line after will format it as a blockquote if that's what you're going for.

Comment (mlarcher) [inline docs/rules/id-match.md]: Thanks, adding it right now

**Comment by mlarcher** [comment]
I've fixed a glitch in the doc and replace "always"/"never" with "yes"/"bi" which I find more appropriate.
Also, I'm not sure about the inline exclusion in my example on https://github.com/mlarcher/eslint/blob/regex-match-identifier/docs/rules/id-match.md , could someone tell me if that looks alright ?

File: docs/rules/id-match.md [hunk 3]
```diff
@@ -0,0 +1,73 @@
+# Require IdMatch (id-match)
```
Comment (nzakas) [inline docs/rules/id-match.md]: Require IDs to match a pattern

**Comment by nzakas** [comment]
I'd use true/false booleans instead of "yes"/" no"

**Comment by nzakas** [comment]
What do you mean by inline exclusion?

**Comment by mlarcher** [comment]
I thought I had to restrict myself to strings, I'll update to booleans.

I meant `/*eslint id-match: [2, "", {checkProperties: "no"}]*/`

**Comment by mlarcher** [comment]
I've replaced strings with booleans and inversed the logic (checkProperties became ignoreProperties) so that I can have the correct default if nothing is set and allow any truthy/falsy value.

**Comment by nzakas** [comment]
We typically don't use negating words in options. I think calling the option `properties` with a default value of false is the way to go.

Also, keep in mind that MemberExpressions also match reads, so properties are a bit more complicated. What if you are just consuming a property you didn't create? Like:

```js
var x = obj._foo;
```

If you don't own `obj`, then warning on this is strange.

**Comment by mlarcher** [comment]
Ok for the positivity ;)
I've use `checkProperties` instead of `properties` though, as I find it to be more explicit.

I also added a test for the case you highlighted.

**Comment by nzakas** [comment]
Sorry to nit, but "check" is completely redundant and doesn't fit with the way our rule options are typically phrased. Consistency is important when you have almost 200 rules.

**Comment by mlarcher** [comment]
@nzakas you got me there. I am a huge fan of explicity, but nothing can beat consistency. I reverted again to "properties" in the options.

**Comment by nzakas** [comment]
Thanks for your patience. This looks good to go!

**Comment by mlarcher** [comment]
Sweet :) Can't wait to get it from the next release and use it on my projects! Any idea when that will happen ?

**Comment by nzakas** [comment]
Hopefully this week

**Pull request merged**